walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,68 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(final Collection<Event> events, final MeetingRequest request) {
+    if (events.isEmpty() && request.getAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    List<TimeRange> impossbleTimeRanges = new ArrayList<>();
+    events.forEach(e -> {
+      Set<String> attendeeIntersection = new HashSet<>(e.getAttendees());
+      attendeeIntersection.retainAll(request.getAttendees());
+      if (attendeeIntersection.size() != 0) {
+        impossbleTimeRanges.add(e.getWhen());
+      }
+    });
+    Collections.sort(impossbleTimeRanges, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> freeTimeRanges = findFreeTimeRangesIn(impossbleTimeRanges, TimeRange.START_OF_DAY,
+        TimeRange.END_OF_DAY);
+    List<TimeRange> result = freeTimeRanges.stream().filter(t -> t.duration() >= request.getDuration())
```

Comment:
Nit: this can be returned directly:
```
return freeTimeRanges.stream().filter(t -> t.duration() >= request.getDuration()).collect(Collectors.toList());
```